### DISCUSSION TITLE
Adjusting 2 sequential 'if' conditions in a row using blocks

### DISCUSSION THREAD

**Discussion by EmperorKabir** [pybricks/support#1710] [Q&A]
Using Claude AI, I've been trying to understand how sequential 'if' conditons work. Here is a sample bit of code which I've butchered from another bit of code (hence you'll notice some unused blocks):

![image](https://github.com/user-attachments/assets/db14ba6a-6079-46e8-8337-e1610392e6ff)

To my untrained mind, if for example 'left +' is pressed, the code will essentially end up getting to a first 'if' condition confirming that left+ is pressed and then when moving to the 'Button Reversion' function, immediately move to the next line where it will again check an identical 'if' condition.

Claude AI suggests that in real world application, a human pressing a button quickly is far slower than the polling rate of the lego remote and so it will confirm both if conditions as true, ultimately setting the power at 70% from what the human perspective is a quick tap of the left+ button. In other words, the human sees/feels it as pressing the button once, Python sees it as being pressed once and then once again due to having not been depressed.

My questions are:

-am I right in this reasoning?

-If so, what is the real world polling rate when using a lego remote, a main city hub (and also a second city hub that mimics the first city hub by receiving bluetooth commands from the first city hub)? - In other words, what is the value of the slowest polling rate between: the lego remote itself, the bluetooth between the remote and the hub, the bluetooth between the hub and the second hub etc etc. I ask in case of the risk of the human potentially being able to depress the button faster than the second if condition running (or causing that to happen some other way), causing the code to call the button reversion function but not begin executing anything within it i.e. first if condition met but second one not being met.

-If for argument's sake I wanted to enforce a break between the if conditions, do I simply input a 'wait until button not pressed' block or is there more to consider than that?

(This is all being asked due to various bugs in a bigger lot of block code I have).

**Comment by laurensvalk** [comment]
> -am I right in this reasoning?

Polling speed isn't really an issue, but I'm not sure the `Button Reversion` block is quite right. It will only run if `left +` or `left -` are pressed. Maybe did you mean to put it in the else part of the switch under your main start program block?

>  In other words, what is the value of the slowest polling rate between: the lego remote itself, the bluetooth between the remote and the hub, the bluetooth between the hub and the second hub etc etc. I ask in case of the risk of the human potentially being able to depress the button faster than the second if condition

The hub and remote are generally faster, but wireless communication could sometimes be a bit slower than you releasing the button.

That isn't necessarily an issue if you make sure your program doesn't rely on exact timing. The communication to the other hub shouldn't have to be the bottleneck here, I think. If you manage all your button logic on the main hub and just send a required power value to the other hub, it doesn't depend on wireless communication or button press speed.

**Reply by EmperorKabir** [reply]
Thanks for the quick response! The button reversion is a bit of butchered code - I'm unsure if you are asking if this is fundamentally broken or because it is missing the larger code; my main program is as follows; I'm trying to run various different functions depending on what button is pressed but the complexity lies in wanting the code to immediately break out of whatever function is running, regardless of what stage it is in within the function and immediately move to the function associated with that particular button.

It half works but it seems quite buggy so I've been trying to understand how to refine this. The repeated if functions/wait functions are my somewhat confused attempts at getting it to work:

![image](https://github.com/user-attachments/assets/0a736fcb-b656-4ee7-ac10-a0a26d203973)
